Implement disambiguation rules.

ltratt opened the pull request:
**Don't merge yet: comments only!**

ltratt commented:
The first part of this PR implements Yacc's default disambiguation rules. Personally, I think these are bonkers, and the default should be to throw an error. But if Yacc does it, and we want to be Yacc compatible, this is what we have to do.

@ptersilie I'd be grateful if you could see if the tests in the first commit look semi-sensible or not.

ptersilie commented:
I think the main reason Yacc does this is probably for the dangling else problem which I'm not sure how to solve properly without this.

ltratt commented:
OK, fair enough.

ptersilie commented:
Tests look ok.

ltratt commented:
OK, thanks!

ltratt commented:
[This PR is so old, I forgot it even existed. I force pushed without even thinking about it. Mea culpa!]

OK, this now (hopefully) fully implements left and right precedences. It doesn't report conflicts to the user, though it does count them; I think we should leave reporting to a later PR (working out a sensible API for this won't be trivial).

ltratt commented:
Hmm, just noticed that the PHP grammar uses `%nonassoc` so I should probably implement that too. Joy.

ltratt commented:
OK, support for `%nonassoc` is now in there, so I think this ready for review by @ptersilie.

ptersilie commented:
My rust knowledge is way to rusty (no pun intended) to be able to say anything about the implementation, but the tests look alright, so if they succeed this should be fine.

Maybe it's worth having @jaheba have a look at the implementation.

ltratt commented:
Good idea. I'll add him as a reviewer.

jaheba commented on src/lib/grammar.rs:
```diff
@@ -115,25 +132,33 @@ pub fn ast_to_grammar(ast: &grammar_ast::GrammarAST) -> Grammar {
     }
 
-    let mut prods = Vec::new();
+    let mut prods                               = Vec::new();
```
alignment

ltratt replied:
This lines up with line below.

jaheba commented on src/lib/statetable.rs:
```diff
@@ -1,14 +1,18 @@
 pub struct StateTable {
-    pub actions: HashMap<(usize, Symbol), Action>,
-    pub gotos  : HashMap<(usize, PIdx), usize>
+    pub actions      : HashMap<(usize, Symbol), Action>,
```
alignment

ltratt replied:
Ditto.

jaheba replied:
Yes, but why?

On Friday, 15 July 2016, Laurence Tratt <notifications@github.com> wrote:

> In src/lib/statetable.rs
> <https://github.com/softdevteam/lrtable/pull/37#discussion_r70982103>:
>
> >  use stategraph::StateGraph;
> >
> >  /// A representation of a StateTable for a grammar. `actions` and `gotos` are split into two
> >  /// separate hashmaps, rather than a single table, due to the different types of their values.
> >  #[derive(Debug)]
> >  pub struct StateTable {
> > -    pub actions: HashMap<(usize, Symbol), Action>,
> > -    pub gotos  : HashMap<(usize, PIdx), usize>
> > +    pub actions      : HashMap<(usize, Symbol), Action>,
>
> Ditto.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/softdevteam/lrtable/pull/37/files/0f7fb8c6ca478869042c7d04d4eea929cb7d9a73#r70982103>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIKZGwl-4Z5xDgzWlh-YI5EB38xyCRbks5qV5lSgaJpZM4Hlhr5>
> .
>

jaheba commented on src/lib/statetable.rs:
```diff
@@ -64,14 +76,64 @@ impl StateTable {
+                                                            // favour of the reduce (i.e. leave
+                                                            // as-is).
+                                                        }
+                                                        (AssocKind::Right, AssocKind::Right) => {
```
inconsistent `,` after match block

jaheba commented on src/lib/statetable.rs:
```diff
@@ -64,14 +76,64 @@ impl StateTable {
+                                                        }
+                                                    }
+                                                }
+                                                else if term_lev > prod_lev {
```
else belongs on the same line as the closing `}`

jaheba commented on src/lib/yacc.rs:
```diff
@@ -51,14 +54,13 @@ impl fmt::Display for YaccError {
     fn new(src: String) -> YaccParser {
         YaccParser {
-            src: src,
+            src:     src,
```
alignment

jaheba commented on src/lib/yacc.rs:
```diff
@@ -90,7 +93,36 @@ impl YaccParser {
+                    k = j;
+                }
+                else if let Some(j) = self.lookahead_is("%right", i) {
+                    kind = AssocKind::Right;
```
no new line

jaheba commented:
There are some things in general in the code base which I would consider not in line with the rust style of life. But I would suggest to merge this PR and then do the styling in a second step.

These include:

* `String::from` instead of `to_string`. Also `to_owned` is better than `to_string` for `&str`.
* Trailing commas. The last line in structs/enums and match blocks should also end with an `,`
* Also consistent usage of `,` after `}` in match blocks.
* No whitespace alignment for assignments, etc.
* Use `for x in &y` instead of `for x in y.iter()`.
* `} else ...` instead of `} \n else ...`
* `&String to &str`
* `&String == &str` to `String == &str`
* `match *foo { bar => ... ` instead of `match foo { &bar =>`

ltratt commented:
I agree that this repo needs reconsidering in light of the many changes to Rust since we started, and that we should do that in a new PR.

ltratt commented:
@ptersilie Question: can I ever end up trying to resolve a shift/reduce where the two kinds aren't the same? (e.g. the terminal is Left associative but the non-term is Right associative) I feel that this can probably happen, but I haven't managed to trigger it yet...

ltratt commented:
@jaheba I've addressed the comments I agreed with (I didn't agree with the alignment ones, as you probably realised ;) ). Unfortunately they're not easily squashable, because they relate to several other commits, so I'm not entirely sure what the best thing is to do. [I don't want to squash this PR to one commit, as it's too hard to understand that way.]

jaheba commented on src/lib/statetable.rs:
```diff
@@ -148,6 +101,54 @@ impl StateTable {
+    let term_k_prec = grm.terminal_precs[term_k];
+    let prod_k_prec = grm.prod_precs[prod_k];
+    match (term_k_prec, prod_k_prec) {
+        (None, None) | (Some(_), None) | (None, Some(_)) => {
```
Can't this be simplified to `(None, _) | (_, None)`?

ltratt replied:
Yes, good spot!

ptersilie commented:
@ltratt I'm not sure. I have to admit implementing this stuff is quite a while ago so I don't remember all the details. Isn't there maybe some rule that, say, prioritises terminal pecedence over nonterminal precedence?

ptersilie commented:
It might also be that there is a natural precedence caused by how the parser looks at things, i.e. the parser sees the terminal first so uses it's left association so doesn't even get to see that the Nonterminal also has a precedence. We would need some actual example though to properly think about this.

jaheba commented on src/lib/statetable.rs:
```diff
@@ -148,6 +101,54 @@ impl StateTable {
+            shift_reduce += 1;
+        },
+        (Some(Precedence{level: term_lev, kind: term_kind}),
+         Some(Precedence{level: prod_lev, kind: prod_kind})) => {
```
This could be simplified to `(Some(term@Precedence{..}), Some(prod@Precedence{..}))`
In the body `term.level`, etc are then available.

ltratt replied:
Good idea.

jaheba commented:
> I didn't agree with the alignment ones, as you probably realised ;)

Understood, but just for reference this isn't normally done in [rust](https://github.com/rust-lang/rust/blob/master/src/doc/style/style/whitespace.md#alignment)

ltratt commented:
@ptersilie At the moment, things panic if the two kinds aren't the same. So maybe we'll just have to wait until an example occurs in real life... unless we can think of one before!

ltratt commented:
If you're OK with the fixes, I will squash (to the extent that I can).

jaheba commented:
@ltratt Agreed :)

ltratt commented:
OK, assuming the tests pass, I think this ready for merging.

ltratt commented:
Perhaps inevitably, Travis fails to compile things (the regex library?!), but everything runs tickety-boo on my local machine.

ltratt commented:
Ah, I think this is because travis is using rust nightly (and OpenBSD currently has 1.9.0, whereas the latest stable as of 11 days ago is 1.10.0).

ltratt commented:
OK, so the problem was the Rust nightly on Travis. With that, everything is fixed (and, although it wasn't strictly necessary, using a more recent version of the regex library is probably a good idea).

So, if you guys are happy with this, please feel free to merge. [Note that I managed to split the "addresses jasper's comment" into bits and squash them appropriately. I felt very pleased with myself!]
